### PR TITLE
crimson: make seastar::do_with() a friend of errorated futures.

### PR DESCRIPTION
**Pull request by rzarzynski**
Signed-off-by: Radoslaw Zarzynski <rzarzyns@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/common/errorator.h [hunk 1]
```diff
@@ -604,6 +604,10 @@ struct errorator {
 
     template <class...>
     friend class ::seastar::future;
+
+    // let seastar::do_with to up-cast us to seastar::future.
+    template<typename T, typename F>
+    friend inline auto ::seastar::do_with(T&&, F&&);
```
Comment (tchaikov) [inline src/crimson/common/errorator.h]: i don't really understand this. by making `::seastar::do_with(T&&, F&&)`, this function still cannot cast `_future` to its private base class.

Comment (rzarzynski) [reply src/crimson/common/errorator.h]: `seastar::do_with()` internally calls `seastar::internal::set_callback()` which takes `seastar::future` as it first parameter. Making `do_with` our friend actually allows for the implicit up-cast conversion to take place there.

Comment (tchaikov) [reply src/crimson/common/errorator.h]:
```c++
struct foo_t {
  int i;
};

struct bar_t : private foo_t {
  bar_t() = default;
private:
  friend void func(foo_t&);
};

void func(foo_t& f) {
  
}

int main()
{
  bar_t b;
  func(b);
}
```

does not compile. see https://godbolt.org/z/QNStKP

Comment (rzarzynski) [reply src/crimson/common/errorator.h]: That's not the case. Let me mimic `do_with` with the following snippet:

```cpp
struct foo_t {
  int i;
};

struct bar_t : private foo_t {
  bar_t() = default;
private:
  friend void barfunc(bar_t&);
};

void foofunc(foo_t& f) {

}
void barfunc(bar_t& f) {
  foofunc(f);
}

int main()
{
  bar_t b;
  barfunc(b);
}
```

https://godbolt.org/z/xbgLas

**Pull request merged**